sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala

```diff
@@ -51,8 +51,10 @@ class SparkOptimizer(
     Batch("Optimize Metadata Only Query", Once, OptimizeMetadataOnlyQuery(catalog)) :+
     Batch("PartitionPruning", Once,
       PartitionPruning,
-      RowLevelOperationRuntimeGroupFiltering,
-      OptimizeSubqueries) :+
+      // We can't run `OptimizeSubqueries` in this batch, as it will optimize the subqueries
+      // twice which may break some optimizer rules that can only be applied once. The rule below
+      // only invokes `OptimizeSubqueries` to optimize newly added subqueries.
```

Comment:
Hm? This batch has only `PartitionPruning` and `RowLevelOperationRuntimeGroupFiltering`. What some optimizer rules are? `PartitionPruning`?

Reply:
Oh, you mean other Once batches in `SparkOptimizer.defaultBatches`?

Reply:
But in `Optimizer` where `OptimizeSubqueries` also runs, there are also other Once batches but seems fine?

Reply:
All the optimizer batches are optimizing the same query plan. If `OptimizeSubqueries` appears twice, it means the subqueries are optimized twice.

Note that, most optimizer rules don't optimize subqueries, they need `OptimizeSubqueries` to invoke the entire optimizer to optimize subqueries recursively.

Reply:
inspired by https://github.com/apache/spark/pull/38619 , maybe we don't need to invoke the entire optimizer, but just a few rules to optimize this subquery.

Reply:
> All the optimizer batches are optimizing the same query plan. If `OptimizeSubqueries` appears twice, it means the subqueries are optimized twice.

Oh, got it, you actually mean `OptimizeSubqueries` is applied twice (here and `Optimizer`). I thought that by running `OptimizeSubqueries` itself here breaks some rules which cannot run twice.